src/ocean/meta/traits/Basic.d

```diff
@@ -0,0 +1,424 @@
+    static assert ( isIntegerType!(ubyte));
+    static assert (!isIntegerType!(char));
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is a floating point type, excluding complex/imaginary ones
+
+*******************************************************************************/
+
+public template isRealType ( T )
+{
+    const bool isRealType =
+           is( Unqual!(T) == float )
+        || is( Unqual!(T) == double )
+        || is( Unqual!(T) == real );
+}
+
+///
+unittest
+{
+    static assert ( isRealType!(double));
+    static assert (!isRealType!(long));
+    static assert (!isRealType!(cdouble));
+}
+
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is a complex floating point type
+
+*******************************************************************************/
+
+public template isComplexType( T )
+{
+    const bool isComplexType =
+           is( Unqual!(T) == cfloat )
+        || is( Unqual!(T) == cdouble )
+        || is( Unqual!(T) == creal );
+}
+
+///
+unittest
+{
+    static assert ( isComplexType!(cdouble));
+    static assert ( isComplexType!(Const!(cdouble)));
+    static assert (!isComplexType!(double));
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is an imaginary floating point type
+
+*******************************************************************************/
+
+public template isImaginaryType( T )
+{
+    const bool isImaginaryType =
+           is( Unqual!(T) == ifloat )
+        || is( Unqual!(T) == idouble )
+        || is( Unqual!(T) == ireal );
+}
+
+///
+unittest
+{
+    static assert ( isImaginaryType!(idouble));
+    static assert ( isImaginaryType!(Const!(idouble)));
+    static assert (!isImaginaryType!(double));
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is any floating-point type: real, complex, or
+        imaginary.
+
+*******************************************************************************/
+
+public template isFloatingPointType( T )
+{
+    const bool isFloatingPointType =
+           isRealType!(T)
+        || isComplexType!(T)
+        || isImaginaryType!(T);
+}
+
+///
+unittest
+{
+    static assert (isFloatingPointType!(double));
+    static assert (isFloatingPointType!(ifloat));
+}
+
+/*******************************************************************************
+
+    Used by other traits to distinguish between dynamic and static arrays
+    instead of plain `bool` values.
+
+*******************************************************************************/
+
+public enum ArrayKind
+{
+    NotArray    = 0,
```

Comment:
Maybe note: `// defined as 0 so that the value of isArrayType!() can be used as a bool`.